components/browser/icons/src/main/java/mozilla/components/browser/icons/loader/NonBlockingHttpIconLoader.kt

```diff
@@ -0,0 +1,40 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.icons.loader
+
+import android.content.Context
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import mozilla.components.browser.icons.IconRequest
+import mozilla.components.concept.fetch.Client
+
+/**
+ * [HttpIconLoader] variation that will immediately resolve a [load] request with [IconLoader.Result.NoResult]
+ * and then continue to actually download the icon in the background finally calling [loadCallback]
+ * with the actual result and details about the request.
+ *
+ * @property httpClient [Client] used for downloading the icon.
+ * @property scope [CoroutineScope] used for downloading the icon in the background.
+ * @property loadCallback Callback for when the network icon finished downloading or an error or timeout occurred.
+ */
+class NonBlockingHttpIconLoader(
+    httpClient: Client,
+    private val scope: CoroutineScope,
+    private val loadCallback: (IconRequest, IconRequest.Resource, IconLoader.Result) -> Unit
+) : HttpIconLoader(httpClient) {
+    override fun load(context: Context, request: IconRequest, resource: IconRequest.Resource): IconLoader.Result {
+        if (!shouldDownload(resource)) {
+            return IconLoader.Result.NoResult
+        }
+
+        scope.launch {
```

Comment:
I wonder if we need a different scope here? Otherwise we'd still keep the icon threads busy 🤔

Reply:
Dispatchers.IO as suggested offline seems great, thank you!

Reply:
Went with having `BrowserIcons` pass a scope with a larger thread pool.
But maybe it would be better for this loader to hardcode `Dispatcher.IO`?

Reply:
Yeah, I think hard-coding `Dispatcher.IO` is fine. The parameter may make sense for testing though?

Reply:
Having `CoroutineScope(Dispatcher.IO)` as the default argument seems like the best compromise.